### DISCUSSION TITLE
Project metadata in pyproject.toml vs setup.cfg

### DISCUSSION THREAD

**Discussion by mwtoews** [pypa/cibuildwheel#1184] [Q&A] [answered]
I've been seeking to move static metadata for a python project from `setup.py` to either `setup.cfg` or `pyproject.toml` ([PEP 621](https://peps.python.org/pep-0621/)), but have found that cibuildwheel does not treat these equally. Consider two pull requests for [shapely](https://github.com/shapely/shapely):

1. https://github.com/shapely/shapely/pull/1426
2. https://github.com/shapely/shapely/pull/1430

This project uses [`release.yml`](https://github.com/shapely/shapely/blob/883af5af268c08cb4921075eee74be5ef1d00730/.github/workflows/release.yml) to automate binary wheels using cibuildwheel, and there have been no issues obtaining project metadata from `setup.py`. There also seems to be no issues with the second pull request using `setup.cfg`.

With the `pyproject.toml` approach, the trouble is in the CI section "Building wheel...":
```
      + rm -rf /tmp/cibuildwheel/built_wheel
      + mkdir -p /tmp/cibuildwheel/built_wheel
      + python -m pip wheel /project --wheel-dir=/tmp/cibuildwheel/built_wheel --no-deps
  Processing /project
    Installing build dependencies: started
    Installing build dependencies: finished with status 'done'
    Getting requirements to build wheel: started
    Getting requirements to build wheel: finished with status 'done'
    Installing backend dependencies: started
    Installing backend dependencies: finished with status 'done'
    Preparing metadata (pyproject.toml): started
    Preparing metadata (pyproject.toml): finished with status 'done'
  Building wheels for collected packages: UNKNOWN
    Building wheel for UNKNOWN (pyproject.toml): started
    Building wheel for UNKNOWN (pyproject.toml): finished with status 'done'
    Created wheel for UNKNOWN: filename=UNKNOWN-2.0.0.dev0+84.gb94ae9a-cp36-cp36m-linux_x86_64.whl size=769612 sha256=8f4b9965fb678844f7c053512b03c10074d9bbacd98c76db24b4485699b2ee47
    Stored in directory: /tmp/pip-ephem-wheel-cache-pww8hbhd/wheels/2e/83/83/fb24504ded536a3c91dd77593aa9037a98b0009fb86c7ebcc0
  Successfully built UNKNOWN
      + /opt/python/cp38-cp38/bin/python -c 'import sys, json, glob; json.dump(glob.glob('"'"'/tmp/cibuildwheel/built_wheel/*.whl'"'"'), sys.stdout)'
      + rm -rf /tmp/cibuildwheel/repaired_wheel
      + mkdir -p /tmp/cibuildwheel/repaired_wheel
```
It seems that one of the components is unable to extract the project metadata for the project name "shapely" from `pyproject.toml`, which is why "UNKNOWN" is shown. The dynamic data for version "2.0.0.dev0+84.gb94ae9a" was successfully obtained from `setup.py` (using versioneer).

Questions:

- What component is unable to get the project metadata for the project name? Is it PEP 621 compatible? If not, is there a plan to implement the feature?
- Are there obvious user errors that are blocking project metadata from `pyproject.toml`?
- There are no issues with `setup.cfg`. Is this because these components are mature?
- If there is a better forum to ask for help, please redirect me.

**Comment by henryiii** [comment, marked as answer]
You need setuptools 61+ for pyproject.toml. But that version doesn’t support python 3.6.

It’s the job of your build backend, not cibuildwheel, FYI. We only read Python requires and our own config. :)

**Reply by mwtoews** [reply]
It seems that wheels for 3.6 cannot be built; this appears to be the fix `CIBW_SKIP: cp36-*`
